ibstore/_minimize.py

```diff
@@ -114,6 +114,27 @@ def _run_openmm(
             force_field_name=input.force_field,
         )
 
+    elif input.force_field.startswith("espaloma"):
+        import espaloma as esp
+
+        # espaloma needs an OpenFF force field too, so provide it as
+        # espaloma-openff-2.1.0, for example. use a slice in case it's empty
+        ff = input.force_field.split("-", 1)[1:2]
```

Comment:
What do you envision the inputs looking like here? I interpret this to expect i.e. `espaloma-openff-3.0.0.offxml`

Reply:
That's what I thought at first too, but Espaloma always adds `.offxml`, see [here](https://github.com/choderalab/espaloma/blob/13af90b8158d50730c2c6888dacce8351021f797/espaloma/graphs/deploy.py#L32). So it has to be something like `espaloma-openff-2.1.0` exactly. I'm also not married to separating the parts by dashes and splitting the string if you have a better idea. It was just the first thing that came to mind.

Reply:
I'm pretty confused. I thought Espaloma was meant to be a replacement for force fields, so it's surprising to see that it requires one at all.

Looking at the code, it seems like it [uses the toolkit](https://github.com/choderalab/espaloma/blob/13af90b8158d50730c2c6888dacce8351021f797/espaloma/graphs/deploy.py#L89-L109) for assigning charges using non-NN methods, but doesn't for NN-based charge assignment. Okay, that makes sense locally. The rest of the logic, though, seems to involve replacing every single result of SMIRNOFF-based parameterization with NN-derived values. I guess this makes sense from the perspective of keeping things structured with a similar feel, but much of the CPU time spent looking up parameters and assigning them to the `openmm.Force`s is wasted. If I'm looking at the logic correctly, it only uses the vdW values?

But, back to my original confusion. Espaloma is assining its own charges, bond parameters, angle parameters, and torsion parameters, leaving only vdW parameters derived from the SMIRNOFF force field. So force fields with the same vdW parameters (say - all of Parsley, or all of Sage) should produce redundant results from different `f"espaloma-{ff}"` tags. Similarly, experiments refitting any or all valence parameters but not vdW terms will also be redundant.

Since Espaloma isn't really a force field, nor is it entirely using one, it might be better to have a separate field to handle this.

Reply:
I agree with the assessment of Espaloma's internal structure as strange. What you say about the redundant results also makes sense. I guess I should have opened an issue first.

Basically I just want a single Espaloma benchmark to compare to some of my SMIRNOFF force fields. All of the other benchmarks are generated with ibstore, so I wanted the Espaloma benchmark to be produced in the same way.

My first draft of this commit didn't specify a SMIRNOFF force field (i.e. it used the default `openff_unconstrained-2.0.0` from espaloma) and my test run crashed for some issue assigning parameters. When I instead passed `openff_unconstrained-2.1.0` (as I had done in my labeling tests) it worked successfully. So I really only care about running a single Espaloma benchmark where "forcefield" is `espaloma-openff_unconstrained-2.1.0`.

Should I just monkey-patch `_run_openmm` for this single benchmark instead of adding this code to the package itself? I thought this functionality would be more generally useful, but maybe only for people attempting to run the same single benchmark. Although there could be multiple espaloma benchmarks if we make the espaloma model (currently hard-coded to `latest`) a parameter too.

If you do want to keep something like this change, adding another field is fine with me. I just thought this would keep the change to a minimum and take advantage of `forcefield` already being passed around everywhere.

Reply:
Heh, well, if Espaloma crashes with the constrained versions of Sage that seems like an external problem.

I think this framework being adaptable to Espaloma and other force field-adjacent tools would be beneficial, even if it's supposed to be something more like a plugin. Do you think you could move around your change to follow the pattern [here](https://github.com/mattwthompson/ib/blob/main/ibstore/_forcefields.py#L32-L65)? Looking at it now, this is how I intended for something to work at some point in time, but details are fuzzy and I don't think it's really used in the way I originally envisioned. (You could imagine this expanding to also cover Foyer, non-GAFF XMLs provided by OpenMM, or even stuff like OPLS via a plug-in.) I don't think `_get_force_field_type` is used now - it's fine to branch out [here](https://github.com/mattwthompson/ib/pull/6/files#diff-ab72fc0662bb971a1fcbfa7032bbd72f58b17d6b89fae08c009f63f6797638a0R117) as you currently do and that can be refactored later in one big sweep.

I don't love squishing a force field name after `espaloma-` but if you can get something working for now without needing to rewrite unrelated code, I think that's great.

Reply:
Yeah I almost mentioned that the crash might be a bug in Espaloma! But I just tried to work around that for now.

Sure thing, I'll move the majority of the code to an `_espaloma` function more like the GAFF case.

Thanks for the feedback and help!

Reply:
Espaloma assumes all (topological) bonds have corresponding parameters in the `openmm.HarmonicBondForce` but Interchange does not add these parameters to bonds (or angles) between atoms that are constrained. Espaloma seems to make this assumption, which is causing a crash that I predict matches what you ran into. I figure they just switched to the unconstrained force field at some point in time and never saw a need to go back, so they're not testing the constrained version. https://github.com/choderalab/espaloma/issues/189